### PR TITLE
[SPARK-5187][SQL] Fix caching of tables with HiveUDFs in the WHERE clause

### PR DESCRIPTION
**Pull request by marmbrus**

### CODE REVIEW

**Comment by cfregly** [comment]
lgtm.  as we just discussed, this is the same code path as SchemaRDD.cache(), so no need for additional tests.

File: sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala [hunk 1]
```diff
@@ -53,7 +53,7 @@ import scala.language.implicitConversions
  *
  * @param functionClassName UDF class name
  */
-class HiveFunctionWrapper(var functionClassName: String) extends java.io.Externalizable {
+case class HiveFunctionWrapper(var functionClassName: String) extends java.io.Externalizable {
```
Comment (sarutak) [inline sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala]: nit: should `functionClassName` be still `var`?

Comment (marmbrus) [inline sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala]: Yeah, its mutated below by our custom deserialization.

Comment (sarutak) [inline sql/hive/v0.13.1/src/main/scala/org/apache/spark/sql/hive/Shim13.scala]: Ah, I see. It's externalaizable.

**Comment by marmbrus** [comment]
Merged to master and branch-1.2

**Pull request closed without merging**